### ISSUE TITLE
Validate all incoming data before caching or forwarding it

### ISSUE THREAD

**Issue by DemiMarie** [bluejekyll/trust-dns#346]
trust-dns should validate all incoming data before caching or forwarding it.  That means:

- checking that the data is syntactically correct
- checking that the data is semantically well-formed
- checking DNSSEC signatures, if enabled

Furthermore, trust-dns should drop all non-conforming input, and ensure that all output is conforming.  This helps prevent attacks as CVE-2015-7547 (glibc getaddrinfo RCE) when trust-dns is used e.g. as the DNS server for a LAN.

**Comment by bluejekyll** [comment]
Is it the server which is not currently doing this? I believe both the client and resolver meet all of these requirements.

I haven't yet built any forwarding features into the server.

**Comment by DemiMarie** [comment]
This (along with the other two issues I just raised) are more abstract principles, but yes I was referring to the Server.  I was hoping that trust-dns could be used as a replacement for the insecure BIND and for djbdns, which is very secure (only one vulnerability ever discovered, and that only affected 32-bit systems), but is unmaintained (requiring 3rd party patches to support IPv6 and DNSSEC).

**Comment by bluejekyll** [comment]
Yes, this is a goal of mine as well. I have plans to build forwarding features into the Server, by way of incorporating the Resolver. I haven't started on that work. As an authority though, I think it meets  the goals listed above.

The Resolver also implements all of those, so once forwarding is implemented, then a forwarding server should meet all of these requirements.
